File: rust/0020_valid_parentheses.rs
//https://leetcode.com/problems/valid-parentheses/submissions/1037086407/

impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.len() % 2 == 1 { return false }

        let mut stack = vec![];
        for c in s.chars() {
            match c {
                '(' => stack.push(')'),
                '{' => stack.push('}'),
                '[' => stack.push(']'),
                ')' | '}' | ']' if Some(c) != stack.pop() => return false ,
                _ => ()
            }
        }
        stack.is_empty()
    }
}

File: java/0020_valid_parentheses.java
/*
 * https://leetcode.com/problems/valid-parentheses/submissions/1286763643/
 */

class Solution {
    public boolean isValid(String s) {
        int N = s.length();
        int top = -1;
        char[] stack = new char[N];
        for (int i = 0; i < N; i++) {
            char c = s.charAt(i);
            if (c == '(') {
                stack[++top] = ')';
            } else if (c == '{') {
                stack[++top] = '}';
            } else if (c == '[') {
                stack[++top] = ']';
            } else if (top != -1 && stack[top] == c) {
                top--;
            } else {
                return false;
            }
        }
        return top == -1;
    }
}